orchagent/neighorch.cpp

```diff
@@ -579,5 +650,438 @@ bool NeighOrch::removeNeighbor(const NeighborEntry &neighborEntry)
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s", op.c_str());
+            it = consumer.m_toSync.erase(it);
+        }
+    }
+}
+
+bool NeighOrch::addInbandNeighbor(string alias, IpAddress ip_address)
+{
+    //For "port" type inband, the inband reachability info syncing can be done through static
+    //configureation or GLOBAL_APP_DB sync (this function)
+
+    //For "vlan" type inband, the inband reachability info syncinng can be ARP learning of other
+    //asics inband or static configuration or through GLOBAL_APP_DB sync (this function)
+
+    //May implement inband rechability info syncing through GLOBAL_APP_DB sync here
+
+    return true;
+}
+
+bool NeighOrch::delInbandNeighbor(string alias, IpAddress ip_address)
+{
+    //Remove inband rechability info sync
+
+    return true;
+}
+
+bool NeighOrch::getSystemPortNeighEncapIndex(string alias, uint32_t &encap_index)
+{
+    string value;
+    if(m_tableVoqSystemNeighTable->hget(alias, "encap_index", value))
+    {
+        encap_index = (uint32_t) stoul(value);
+        return true;
+    }
+    return false;
+}
+
+bool NeighOrch::addVoqEncapIndex(string &alias, IpAddress &ip, vector<sai_attribute_t> &neighbor_attrs)
+{
+    sai_attribute_t attr;
+    uint32_t encap_index = 0;
+    string appKey = alias + ":" + ip.to_string();
+
+    if(getSystemPortNeighEncapIndex(appKey, encap_index))
+    {
+        attr.id = SAI_NEIGHBOR_ENTRY_ATTR_ENCAP_INDEX;
+        attr.value.u32 = encap_index;
+        neighbor_attrs.push_back(attr);
+
+        attr.id = SAI_NEIGHBOR_ENTRY_ATTR_ENCAP_IMPOSE_INDEX;
+        attr.value.booldata = true;
+        neighbor_attrs.push_back(attr);
+    }
+    else
+    {
+        //No Encap index available. Check if remote system port
+        if(gIntfsOrch->isRemoteSystemPortIntf(alias))
+        {
+            //Encap index not available and the interface is remote. Return false to re-try
+            SWSS_LOG_NOTICE("System port neighbor encap index is not available for remote neighbor %s. Re-try!", appKey.c_str());
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool NeighOrch::voqSyncAddNeigh(string &alias, IpAddress &ip_address, const MacAddress &mac, sai_neighbor_entry_t &neighbor_entry)
+{
+    sai_attribute_t attr;
+    sai_status_t status;
+
+    //Sync only local neigh. Confirm for the local neigh and
+    //get the system port alias for key for syncing to GLOBAL_APP_DB
+    Port port;
+    if(gPortsOrch->getPort(alias, port))
+    {
+        if(port.m_system_port_info.type == SAI_SYSTEM_PORT_TYPE_REMOTE)
+        {
+            return true;
+        }
+        alias = port.m_system_port_info.alias;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Port does not exist for %s!", alias.c_str());
+        return false;
+    }
+
+    attr.id = SAI_NEIGHBOR_ENTRY_ATTR_ENCAP_INDEX;
+
+    status = sai_neighbor_api->get_neighbor_entry_attribute(&neighbor_entry, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
```

Comment:
we we need to check encap_index > 0 here ? I see that assumption in some of below code.

Reply:
Yes. Fixed